File: eBPF_Supermarket/Subsystem_Observability/memory/stat.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
#include <linux/mmzone.h>
#include <linux/compaction.h>
#include <linux/mmzone.h>
#include <linux/nodemask.h> 
#include <linux/mm_types.h>
#include "stat.h"

BPF_RINGBUF_OUTPUT(buffer, 1);

struct alloc_context {
        struct zonelist *zonelist;
        nodemask_t *nodemask;
        struct zoneref *preferred_zoneref;
        int migratetype;
        enum zone_type high_zoneidx;
        bool spread_dirty_pages;
};

int sysstat(struct pt_regs *ctx) {
	u64 vm[60];
	struct zoneref *zf; 	
	struct alloc_context *ac;    
        struct pglist_data *pd;

        struct event *event = buffer.ringbuf_reserve(sizeof(struct event));
        if (!event) {
 	      return 1;
        }
	
	ac = (struct alloc_context *)PT_REGS_PARM4(ctx);
        pd = ac->preferred_zoneref->zone->zone_pgdat;
//lru
        bpf_probe_read_kernel(&event->anon_active, sizeof(u64), &pd->vm_stat[1]);
        bpf_probe_read_kernel(&event->anon_inactive, sizeof(u64), &pd->vm_stat[0]);
        bpf_probe_read_kernel(&event->file_inactive, sizeof(u64), &pd->vm_stat[2]);
        bpf_probe_read_kernel(&event->file_active, sizeof(u64), &pd->vm_stat[3]);
        bpf_probe_read_kernel(&event->unevictable, sizeof(u64), &pd->vm_stat[4]);
//working
        bpf_probe_read_kernel(&event->working_nodes, sizeof(u64), &pd->vm_stat[12]);
        bpf_probe_read_kernel(&event->working_refault, sizeof(u64), &pd->vm_stat[13]);
        bpf_probe_read_kernel(&event->working_activate, sizeof(u64), &pd->vm_stat[14]);
        bpf_probe_read_kernel(&event->working_restore, sizeof(u64), &pd->vm_stat[15]);
        bpf_probe_read_kernel(&event->working_nodereclaim, sizeof(u64), &pd->vm_stat[16]);
//page
        bpf_probe_read_kernel(&event->anon_isolated, sizeof(u64), &pd->vm_stat[7]);
	bpf_probe_read_kernel(&event->anon_mapped, sizeof(u64), &pd->vm_stat[17]);
        bpf_probe_read_kernel(&event->file_isolated, sizeof(u64), &pd->vm_stat[8]);
        bpf_probe_read_kernel(&event->file_mapped, sizeof(u64), &pd->vm_stat[18]);

        bpf_probe_read_kernel(&event->shmem, sizeof(u64), &pd->vm_stat[23]);

        bpf_probe_read_kernel(&event->slab_reclaimable, sizeof(u64), &pd->vm_stat[5]);
        bpf_probe_read_kernel(&event->slab_unreclaimable, sizeof(u64), &pd->vm_stat[6]);

	bpf_probe_read_kernel(&event->anon_thps, sizeof(u64), &pd->vm_stat[26]);
	bpf_probe_read_kernel(&event->pmdmapped, sizeof(u64), &pd->vm_stat[25]);

        buffer.ringbuf_submit(event, 0);

    return 0;
}

File: eBPF_Supermarket/Memory_Subsystem/memstat/sysstat/stat.py
from bcc import BPF
import time 
import sys
import os
import argparse

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("-t","--type", default="lru", help="parameter")
    
    return parser.parse_args()

args = parse_args()
b = BPF(src_file="stat.c")
b.attach_kprobe(event_re="get_page_from_freelist", fn_name="sysstat")

def callback(ctx, data, size):
      event = b['buffer'].event(data)
      if args.type == "lru":
            print("%8u %8u %8u %8u %8u %8u %8u" 
                    % (event.anon_active+event.file_active, event.file_active, event.anon_active, event.file_inactive+event.anon_inactive, event.file_inactive, event.anon_inactive, event.unevictable))
      if args.type == "working":
            print("%7u %8u %6u %9u %8u" 
                    % (event.working_nodes, event.working_refault, event.working_activate, event.working_restore, event.working_nodereclaim))
      if args.type == "page":
            print("%8u %11u %10u %11u  --- %4u  --- %6u %9u %13u   --- %4u %11u" 
                    % (event.anon_isolated, event.anon_mapped, event.file_isolated, event.file_mapped,
                        event.shmem, 
                        event.slab_reclaimable+event.slab_unreclaimable, event.slab_reclaimable, event.slab_unreclaimable,
                        event.anon_thps, event.pmdmapped))


b['buffer'].open_ring_buffer(callback)
print("Printing openat() calls, ctrl-c to exit.")

try:
    while 1:
        if args.type == "lru":
            print("%8s %6s %9s %11s %4s %8s %15s" % ("ACTIVE", "FILE","ANON", "INACTIVE", "FILE", "ANON", "UNEVICTABLE"))
        if args.type == "working":
            print("%8s %8s %8s %8s %8s" % ("NODES", "REFAULT", "ACTIVATE", "RESTORE", "NODERECLAIM"))
        if args.type == "page":
            print("ANON:%8s %6s  FILE:%8s %8s %9s %9s %14s %8s %8s %8s" % ("ISOLATED", "MAPPED", "ISOLATED", "MAPPED", "SHMEM", "SLAB", "RECLAIMABLE", "UNRECLAIMABLE", "ANONHUGEPAGE", "SHMEMPMDMAPPED"))
        b.ring_buffer_consume()
        time.sleep(1)
        os.system('clear')

except KeyboardInterrupt:
   sys.exit()
